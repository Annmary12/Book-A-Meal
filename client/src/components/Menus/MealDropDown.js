import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MealCard from '../MealCard/MealCard.js';
import styles from './menus.scss';
import rightArrow from '../../../assets/img/right-arrow.svg';
import downArrow from '../../../assets/img/down-arrow.svg';
import OrderModal from '../OrderModal/OrderModal.js';

/**
 * @class MealDropDown
 *
 */
class MealDropDown extends Component {
  static propTypes = {
    meal: PropTypes.objectOf(PropTypes.oneOfType([
      PropTypes.string, PropTypes.number
    ])).isRequired
  }

  state = {
    droppedDown: false,
    showModal: false
  }

  handleOpenModal = (event) => {
    event.stopPropagation();

    this.setState({
      showModal: true
    });
  }

  handleCloseModal = () => {
    this.setState({
      showModal: false
    });
  }

  dropDown = () => {
    this.setState(prevState => ({ droppedDown: !prevState.droppedDown }));
  }

  /**
   * @returns {JSX} - React JSX
   */
  render() {
    const { meal } = this.props;
    const arrow = this.state.droppedDown ? downArrow : rightArrow;

    return (
      <div className={styles.meal}>
        {/* eslint-disable  */}
        <div className={styles.mealTitle} onClick={this.dropDown}>
        * {meal.title}
          <div className={styles.leftSection}>
            <button className={styles.orderBtn} onClick={this.handleOpenModal}>Order</button>
            <img className={styles.img} src={arrow} alt="down-arrow" />
          </div>
        </div>
        {this.state.droppedDown &&
          <MealCard
            className={styles.mealCard}
            meal={meal}
          />
        }
        <OrderModal
          isOpen={this.state.showModal}
          handleClose={this.handleCloseModal}
          meal={this.props.meal}
        />
      </div>
    );
  }
}

export default MealDropDown;
