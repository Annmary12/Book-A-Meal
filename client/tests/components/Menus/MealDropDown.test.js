import React from 'react';
import MealDropDown from '../../../src/components/Menus/MealDropDown';

const meal = {
  id: 'test-id',
  title: 'fufu and egusi',
  price: 230,
  img: 'fufu.jpg'
};

describe('MealDropDown component', () => {
  const wrapper = shallow(
    <MealDropDown
      meal={meal}
    />
  );
  it('should match snapshot if droppedDown state is true', () => {
    wrapper.instance().setState({ droppedDown: true });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if droppedDown state is false', () => {
    wrapper.instance().setState({ droppedDown: false });

    expect(wrapper).toMatchSnapshot();
  });

  describe('handleOpenOrderModal', () => {
    const event = {
      stopPropagation: jest.fn()
    }
    const { handleOpenOrderModal } = wrapper.instance();
    it('should call stopPropagation on event', () => {
      handleOpenOrderModal(event);

      expect(event.stopPropagation).toBeCalled();
    });

    it('should set showOrderModal to true', () => {
      wrapper.instance().setState({ showOrderModal: false });
      handleOpenOrderModal(event);

      expect(wrapper.instance().state.showOrderModal).toBeTruthy();

    });
  });

   describe('handleCloseOrderModal', () => {
    const { handleCloseOrderModal } = wrapper.instance();

    it('should set showOrderModal to false', () => {
      wrapper.instance().setState({ showOrderModal: true });
      handleCloseOrderModal();

      expect(wrapper.instance().state.showOrderModal).toBeFalsy();
    });
  });

  describe('toggleDropDown method', () => {
    it('should toggle the droppedDown state of the component', () => {
      wrapper.instance().setState({ droppedDown: false});
      wrapper.instance().toggleDropDown();
      expect(wrapper.instance().state.droppedDown).toBeTruthy();
      wrapper.instance().toggleDropDown();
      expect(wrapper.instance().state.droppedDown).toBeFalsy();
    });
  });
});