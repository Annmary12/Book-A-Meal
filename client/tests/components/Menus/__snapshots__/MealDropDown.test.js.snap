// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealDropDown component should match snapshot if droppedDown state is false 1`] = `
<div
  className="meal"
>
  <div
    aria-hidden={true}
    className="mealTitle"
    onClick={[Function]}
    role="button"
  >
    * 
    fufu and egusi
    <div
      className="leftSection"
    >
      <button
        className="orderBtn"
        onClick={[Function]}
      >
        Order
      </button>
      <img
        alt="down-arrow"
        className="img"
        src="test-file-stub"
      />
    </div>
  </div>
  <OrderModal
    handleClose={[Function]}
    isOpen={false}
    meal={
      Object {
        "id": "test-id",
        "img": "fufu.jpg",
        "price": 230,
        "title": "fufu and egusi",
      }
    }
  />
</div>
`;

exports[`MealDropDown component should match snapshot if droppedDown state is true 1`] = `
<div
  className="meal"
>
  <div
    aria-hidden={true}
    className="mealTitle"
    onClick={[Function]}
    role="button"
  >
    * 
    fufu and egusi
    <div
      className="leftSection"
    >
      <button
        className="orderBtn"
        onClick={[Function]}
      >
        Order
      </button>
      <img
        alt="down-arrow"
        className="img"
        src="test-file-stub"
      />
    </div>
  </div>
  <MealCard
    action=""
    className="mealCard"
    deleteMeal={[Function]}
    meal={
      Object {
        "id": "test-id",
        "img": "fufu.jpg",
        "price": 230,
        "title": "fufu and egusi",
      }
    }
  />
  <OrderModal
    handleClose={[Function]}
    isOpen={false}
    meal={
      Object {
        "id": "test-id",
        "img": "fufu.jpg",
        "price": 230,
        "title": "fufu and egusi",
      }
    }
  />
</div>
`;
